### PR TITLE
Added init with QR Code and Code128 to create image of a barcode

### PR DESCRIPTION
**Pull request by heart**
Added init with QR Code and Code128 to create image of a barcode

### CODE REVIEW

File: Tests/UIKitTests/UIImageExtensionsTests.swift [hunk 1]
```diff
@@ -223,6 +223,17 @@ final class UIImageExtensionsTests: XCTestCase {
...
     }
 
-}
+    func testQRCode() {
+        let image = UIImage(qrCode: "qr code is awesome")
+        XCTAssertNotNil(image)
+        XCTAssertNotEqual(image?.size ?? CGSize.zero, CGSize.zero)
```
Comment (guykogus) [inline Tests/UIKitTests/UIImageExtensionsTests.swift:229]: I believe you can check optional values. Test if this works.
```suggestion
        XCTAssertNotEqual(image?.size, .zero)
```

File: Tests/UIKitTests/UIImageExtensionsTests.swift [hunk 2]
```diff
@@ -223,6 +223,17 @@ final class UIImageExtensionsTests: XCTestCase {
...
+    }
 
+    func test2DBarCode() {
+        let image = UIImage(code128Barcode: "code 128 is awesome")
+        XCTAssertNotNil(image)
+        XCTAssertNotEqual(image?.size ?? CGSize.zero, CGSize.zero)
```
Comment (guykogus) [inline Tests/UIKitTests/UIImageExtensionsTests.swift:235]: Same comment as above
```suggestion
        XCTAssertNotEqual(image?.size, .zero)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -14,6 +14,10 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - `center`, `topLeft`, `topRight`, `bottomLeft`, `bottomRight` to get anchor position or set position using anchor. [#835](https://github.com/SwifterSwift/SwifterSwift/pull/835) by [rypyak](https://github.com/rypyak)
 - **CGSize**:
   - Added new `+`, `+=`, `-` and `-=` operator extensions for tuple (width: CGFloat, height: CGFloat). [#841](https://github.com/SwifterSwift/SwifterSwift/pull/841) by [hamtiko](https://github.com/hamtiko)
+- **UIImage**
+  - Added `init?(qrCode:correctionLevel:)` to create QR Code image from string.  [#842](https://github.com/SwifterSwift/SwifterSwift/pull/842) by [heart](https://github.com/heart)
```
Comment (guykogus) [inline CHANGELOG.md:18]:
```suggestion
  - Added `init?(qrCode:correctionLevel:)` to create QR Code image from string.  [#843](https://github.com/SwifterSwift/SwifterSwift/pull/843) by [heart](https://github.com/heart)
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,10 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
...
+  - Added `init?(code128Barcode:quietSpace:)` to create Code 128 barcode image from string.  [#842](https://github.com/SwifterSwift/SwifterSwift/pull/842) by [heart](https://github.com/heart)
```
Comment (guykogus) [inline CHANGELOG.md:19]:
```suggestion
  - Added `init?(code128Barcode:quietSpace:)` to create Code 128 barcode image from string.  [#843](https://github.com/SwifterSwift/SwifterSwift/pull/843) by [heart](https://github.com/heart)
```

File: Tests/UIKitTests/UIImageExtensionsTests.swift (hunk 1 above)
Comment (heart) [reply Tests/UIKitTests/UIImageExtensionsTests.swift:229]: ```XCTAssertNotEqual(image?.size, .zero)```
It's not works It makes the test pass even image size is nil

Comment (guykogus) [reply Tests/UIKitTests/UIImageExtensionsTests.swift:229]: The size will only be `nil` if the image is `nil`, and you already test for that in the line above.

Comment (heart) [reply Tests/UIKitTests/UIImageExtensionsTests.swift:229]: Thank you I got it

**Comment by heart** [comment]
I don't know why SwiftLint failed

**Comment by guykogus** [comment]
@heart you should install SwiftLint yourself to find out why. My guess is because of the very long comment line. Try splitting it up.

**Comment by heart** [comment]
I have tried it on my MacBook
I install SwiftLint from homebrew and run SwiftLint from terminal
but I found 0 violations and 0 serious 

```
swiftlint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths
Linting 'UILayoutPriorityExtensionsTests.swift' (1/207)
Linting 'UIViewControllerExtensionsTests.swift' (2/207)
Linting 'UIColorExtensionsTests.swift' (3/207)
Linting 'UIStackViewExtensionsTests.swift' (4/207)
Linting 'UINavigationBarExtensionTests.swift' (5/207)
Linting 'UILabelExtensionsTests.swift' (6/207)
Linting 'UICollectionViewExtensionsTests.swift' (7/207)
Linting 'UIAlertControllerExtensionsTests.swift' (8/207)
Linting 'UIImageViewExtensionsTests.swift' (9/207)
Linting 'UISegmentedControlExtensionsTests.swift' (10/207)
Linting 'UITableViewExtensionsTests.swift' (11/207)
Linting 'UIScrollViewExtensionsTests.swift' (12/207)
Linting 'UIWindowExtensionsTests.swift' (13/207)
Linting 'UIFontExtensionsTests.swift' (14/207)
Linting 'UIBezierPathExtensionsTests.swift' (15/207)
Linting 'UINavigationControllerExtensionsTests.swift' (16/207)
Linting 'UINavigationItemExtensionsTests.swift' (17/207)
Linting 'UITextFieldExtensionsTests.swift' (18/207)
Linting 'UISearchBarExtensionsTests.swift' (19/207)
Linting 'UIGestureRecognizerExtensionsTests.swift' (20/207)
Linting 'UIStoryboardExtensionsTests.swift' (21/207)
Linting 'UIButtonExtensionsTests.swift' (22/207)
Linting 'UITabBarExtensionsTests.swift' (23/207)
Linting 'UIRefreshControlExntesionsTests.swift' (24/207)
Linting 'UIDatePickerExtensionsTests.swift' (25/207)
Linting 'UITextViewExtensionsTests.swift' (26/207)
Linting 'UIImageExtensionsTests.swift' (27/207)
Linting 'UIViewExtensionsTests.swift' (28/207)
Linting 'UISliderExtensionsTests.swift' (29/207)
Linting 'MKPolylineTests.swift' (30/207)
Linting 'UIBarButtonExtensionsTests.swift' (31/207)
Linting 'UISwitchExtensionsTests.swift' (32/207)
Linting 'MKMapViewTests.swift' (33/207)
Linting 'CalendarExtensionTest.swift' (34/207)
Linting 'NSPredicateExtensionsTests.swift' (35/207)
Linting 'LocaleExtensionsTests.swift' (36/207)
Linting 'URLExtensionsTests.swift' (37/207)
Linting 'NSRegularExpressionExtensionsTests.swift' (38/207)
Linting 'NotificationCenterExtensionsTests.swift' (39/207)
Linting 'DateExtensionsTests.swift' (40/207)
Linting 'UserDefaultsExtensionsTests.swift' (41/207)
Linting 'DataExtensionsTests.swift' (42/207)
Linting 'URLRequestExtensionsTests.swift' (43/207)
Linting 'FileManagerExtensionsTests.swift' (44/207)
Linting 'NSAttributedStringExtensionsTests.swift' (45/207)
Linting 'CAGradientLayerExtensionsTests.swift' (46/207)
Linting 'CATransform3DExtensionsTests.swift' (47/207)
Linting 'SCNPlaneExtensionsTests.swift' (48/207)
Linting 'SCNBoxExtensionsTests.swift' (49/207)
Linting 'SCNMaterialExtensionsTests.swift' (50/207)
Linting 'SCNShapeExtensionsTests.swift' (51/207)
Linting 'SCNConeExtensionsTests.swift' (52/207)
Linting 'SCNGeometryExtensionsTests.swift' (53/207)
Linting 'SCNVector3ExtensionsTests.swift' (54/207)
Linting 'SCNCapsuleExtensionsTests.swift' (55/207)
Linting 'SCNCylinderExtensionsTests.swift' (56/207)
Linting 'SCNSphereExtensionsTests.swift' (57/207)
Linting 'SKNodeExtensionTests.swift' (58/207)
Linting 'CGFloatExtensionsTests.swift' (59/207)
Linting 'CGVectorExtensionsTests.swift' (60/207)
Linting 'CGPointExtensionsTests.swift' (61/207)
Linting 'CGSizeExtensionsTests.swift' (62/207)
Linting 'CGRectExtensionsTests.swift' (63/207)
Linting 'CGColorExtensionsTests.swift' (64/207)
Linting 'BinaryFloatingPointExtensionsTests.swift' (65/207)
Linting 'CGAffineTransformExtensionsTests.swift' (66/207)
Linting 'ArrayExtensionsTests.swift' (67/207)
Linting 'DecodableExtensionsTests.swift' (68/207)
Linting 'DoubleExtensionsTests.swift' (69/207)
Linting 'CharacterExtensionsTests.swift' (70/207)
Linting 'DictionaryExtensionsTests.swift' (71/207)
Linting 'OptionalExtensionsTests.swift' (72/207)
Linting 'SignedIntegerExtensionsTests.swift' (73/207)
Linting 'RandomAccessCollectionExtensionsTests.swift' (74/207)
Linting 'RangeReplaceableCollectionTests.swift' (75/207)
Linting 'StringExtensionsTests.swift' (76/207)
Linting 'BoolExtensionsTests.swift' (77/207)
Linting 'KeyedDecodingContainerExtensionsTests.swift' (78/207)
Linting 'FloatExtensionsTests.swift' (79/207)
Linting 'FloatingPointExtensionsTests.swift' (80/207)
Linting 'IntExtensionsTests.swift' (81/207)
Linting 'TestHelpers.swift' (82/207)
Linting 'SequenceExtensionsTests.swift' (83/207)
Linting 'SignedNumericExtensionsTests.swift' (84/207)
Linting 'StringProtocolExtensionsTests.swift' (85/207)
Linting 'MutableCollectionTests.swift' (86/207)
Linting 'CollectionExtensionsTests.swift' (87/207)
Linting 'ComparableExtensionsTests.swift' (88/207)
Linting 'BidirectionalCollectionExtensionsTests.swift' (89/207)
Linting 'ColorExtensionsTests.swift' (90/207)
Linting 'EdgeInsetsExtensionsTests.swift' (91/207)
Linting 'DispatchQueueExtensionsTests.swift' (92/207)
Linting 'NSImageExtensionsTests.swift' (93/207)
Linting 'NSViewExtensionsTests.swift' (94/207)
Linting 'NSColorExtensionsTests.swift' (95/207)
Linting 'SKProductTests.swift' (96/207)
Linting 'CLLocationArrayExtensionsTests.swift' (97/207)
Linting 'CLLocationExtensionsTests.swift' (98/207)
Linting 'CLVisitExtensionsTests.swift' (99/207)
Linting 'Examples/Examples.playground/Pages/02-FoundationExtensions.xcplaygroundpage/Contents.swift' (100/207)
Linting 'Examples/Examples.playground/Pages/01-SwiftStdlibExtensions.xcplaygroundpage/Contents.swift' (101/207)
Linting 'Examples/Examples.playground/Pages/04-CoreGraphicsExtensions.xcplaygroundpage/Contents.swift' (102/207)
Linting 'Examples/Examples.playground/Pages/05-TryYourself.xcplaygroundpage/Contents.swift' (103/207)
Linting 'Examples/Examples.playground/Pages/03-UIKitExtensions.xcplaygroundpage/Contents.swift' (104/207)
Linting 'Package.swift' (105/207)
Linting 'Examples/Examples.playground/Pages/00-ToC.xcplaygroundpage/Contents.swift' (106/207)
Linting 'SKProductExtensions.swift' (107/207)
Linting 'CGRectExtensions.swift' (108/207)
Linting 'CGFloatExtensions.swift' (109/207)
Linting 'CGVectorExtensions.swift' (110/207)
Linting 'CGSizeExtensions.swift' (111/207)
Linting 'CGAffineTransformExtensions.swift' (112/207)
Linting 'CGPointExtensions.swift' (113/207)
Linting 'CGColorExtensions.swift' (114/207)
Linting 'UIBezierPathExtensions.swift' (115/207)
Linting 'UISearchBarExtensions.swift' (116/207)
Linting 'UITableViewExtensions.swift' (117/207)
Linting 'UIScrollViewExtensions.swift' (118/207)
Linting 'UITextFieldExtensions.swift' (119/207)
Linting 'UILayoutPriorityExtensions.swift' (120/207)
Linting 'UINavigationItemExtensions.swift' (121/207)
Linting 'UIStackViewExtensions.swift' (122/207)
Linting 'UICollectionViewExtensions.swift' (123/207)
Linting 'UITextViewExtensions.swift' (124/207)
Linting 'UIBarButtonItemExtensions.swift' (125/207)
Linting 'UIStoryboardExtensions.swift' (126/207)
Linting 'UITabBarExtensions.swift' (127/207)
Linting 'UISegmentedControlExtensions.swift' (128/207)
Linting 'UINavigationBarExtensions.swift' (129/207)
Linting 'UIActivityExtensions.swift' (130/207)
Linting 'UISliderExtensions.swift' (131/207)
Linting 'UIViewControllerExtensions.swift' (132/207)
Linting 'UIImageExtensions.swift' (133/207)
Linting 'UINavigationControllerExtensions.swift' (134/207)
Linting 'UIViewExtensions.swift' (135/207)
Linting 'UIGestureRecognizerExtensions.swift' (136/207)
Linting 'UIWindowExtensions.swift' (137/207)
Linting 'UILabelExtensions.swift' (138/207)
Linting 'UISwitchExtensions.swift' (139/207)
Linting 'UIButtonExtensions.swift' (140/207)
Linting 'UIApplicationExtensions.swift' (141/207)
Linting 'UIImageViewExtensions.swift' (142/207)
Linting 'UIRefreshControlExtensions.swift' (143/207)
Linting 'UIAlertControllerExtensions.swift' (144/207)
Linting 'UIFontExtensions.swift' (145/207)
Linting 'UIDatePickerExtensions.swift' (146/207)
Linting 'UIColorExtensions.swift' (147/207)
Linting 'DispatchQueueExtensions.swift' (148/207)
Linting 'SCNShapeExtensions.swift' (149/207)
Linting 'SCNSphereExtensions.swift' (150/207)
Linting 'SCNBoxExtensions.swift' (151/207)
Linting 'SCNGeometryExtensions.swift' (152/207)
Linting 'SCNVector3Extensions.swift' (153/207)
Linting 'SCNMaterialExtensions.swift' (154/207)
Linting 'SCNCylinderExtensions.swift' (155/207)
Linting 'SCNPlaneExtensions.swift' (156/207)
Linting 'SCNConeExtensions.swift' (157/207)
Linting 'SCNCapsuleExtensions.swift' (158/207)
Linting 'NSPredicateExtensions.swift' (159/207)
Linting 'NSRegularExpressionExtensions.swift' (160/207)
Linting 'UserDefaultsExtensions.swift' (161/207)
Linting 'DataExtensions.swift' (162/207)
Linting 'URLRequestExtensions.swift' (163/207)
Linting 'FileManagerExtensions.swift' (164/207)
Linting 'NotificationCenterExtensions.swift' (165/207)
Linting 'URLExtensions.swift' (166/207)
Linting 'LocaleExtensions.swift' (167/207)
Linting 'DateExtensions.swift' (168/207)
Linting 'EdgeInsetsExtensions.swift' (169/207)
Linting 'ColorExtensions.swift' (170/207)
Linting 'SKNodeExtensions.swift' (171/207)
Linting 'NSAttributedStringExtensions.swift' (172/207)
Linting 'MKMapViewExtensions.swift' (173/207)
Linting 'CalendarExtensions.swift' (174/207)
Linting 'MKPolylineExtensions.swift' (175/207)
Linting 'SignedIntegerExtensions.swift' (176/207)
Linting 'BidirectionalCollectionExtensions.swift' (177/207)
Linting 'ComparableExtensions.swift' (178/207)
Linting 'ArrayExtensions.swift' (179/207)
Linting 'StringProtocolExtensions.swift' (180/207)
Linting 'BinaryFloatingPointExtensions.swift' (181/207)
Linting 'FloatExtensions.swift' (182/207)
Linting 'KeyedDecodingContainerExtensions.swift' (183/207)
Linting 'SequenceExtensions.swift' (184/207)
Linting 'MutableCollectionExtensions.swift' (185/207)
Linting 'SignedNumericExtensions.swift' (186/207)
Linting 'DictionaryExtensions.swift' (187/207)
Linting 'StringExtensions.swift' (188/207)
Linting 'DecodableExtensions.swift' (189/207)
Linting 'BoolExtensions.swift' (190/207)
Linting 'DoubleExtensions.swift' (191/207)
Linting 'RandomAccessCollectionExtensions.swift' (192/207)
Linting 'CharacterExtensions.swift' (193/207)
Linting 'FloatingPointExtensions.swift' (194/207)
Linting 'IntExtensions.swift' (195/207)
Linting 'RangeReplaceableCollectionExtensions.swift' (196/207)
Linting 'OptionalExtensions.swift' (197/207)
Linting 'CollectionExtensions.swift' (198/207)
Linting 'StdlibDeprecated.swift' (199/207)
Linting 'CLLocationArrayExtensions.swift' (200/207)
Linting 'CLVisitExtensions.swift' (201/207)
Linting 'CLLocationExtensions.swift' (202/207)
Linting 'NSColorExtensions.swift' (203/207)
Linting 'NSViewExtensions.swift' (204/207)
Linting 'NSImageExtensions.swift' (205/207)
Linting 'CAGradientLayerExtensions.swift' (206/207)
Linting 'CATransform3DExtensions.swift' (207/207)
Done linting! Found 0 violations, 0 serious in 207 files.
```

**Comment by heart** [comment]
Still no violations and no serious on Code too

screen shot

**Comment by guykogus** [comment]
Very strange. Move the `#if` conditions back outside the functions for now.
Also, maybe add some tests to see what happens if you use invalid values for `correctionLevel` and `quietSpace`, that should fix the issues for codecov.

**Comment by LucianoPAlmeida** [comment]
Some checks are failing in the checkout action, try to squash the commits in one and send  a force push :)

**Comment by LucianoPAlmeida** [comment]
> Nice and clean, just a small comment.
> @LucianoPAlmeida what do you think regarding our previous conversation in the other PR?

Which comment? Sorry got a bit lost on this 😄

File: Tests/UIKitTests/UIImageExtensionsTests.swift [hunk 5]
```diff
@@ -224,16 +224,26 @@ final class UIImageExtensionsTests: XCTestCase {
...
     func test2DBarCode() {
-        let image = UIImage(code128Barcode: "code 128 is awesome")
+        let image = try? UIImage(code128Barcode: "code 128 is awesome")
         XCTAssertNotNil(image)
         XCTAssertNotEqual(image?.size, .zero)
     }
 
+    func test2DBarCodeWithInvalidQuietSpaceValue() {
+        let image = try? UIImage(code128Barcode: "code 128 is awesome", quietSpace: 900)
+        XCTAssertNil(image)
```
Comment (LucianoPAlmeida) [inline Tests/UIKitTests/UIImageExtensionsTests.swift:245]: I think is possible to use assert throw or something like this ...

**Comment by LucianoPAlmeida** [comment]
This branch needs to be rebased with master to pass the swiftlint build :)

File: Sources/SwifterSwift/UIKit/UIImageExtensions.swift [hunk 6]
```diff
@@ -349,6 +349,52 @@ public extension UIImage {
...
+    ///     -   qrCode: a `String`, representing the QR code image.
+    ///     -   correctionLevel: Controls the amount of additional data encoded to allow larger areas of the code to be damaged.
+    ///  There are four possible correction modes `L` 7% , `M` 15%, `Q` 25%, `H` 30%
+    convenience init?(qrCode: String, correctionLevel: String = "M") throws {
+        if correctionLevel != "L"
+            && correctionLevel != "M"
+            && correctionLevel != "Q"
+            && correctionLevel != "H" {
+            throw NSError(domain: "Invalid correction level. Only 4 possible correction modes L, M, Q, H", code: 0, userInfo: nil)
+        }
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UIImageExtensions.swift:365]: I don't think you need to do this here. At most you can add an `assert`, but don't throw this error (which is wrongly created anyway).

File: Sources/SwifterSwift/UIKit/UIImageExtensions.swift [hunk 7]
```diff
@@ -349,6 +349,52 @@ public extension UIImage {
...
+    #if !os(watchOS)
+    /// SwifterSwift: Create a new QR Code image from a string
+    ///
+    /// - Parameters:
+    ///     -   code128Barcode: a `String`, representing the barcode image.
+    ///     -   quietSpace: The number of pixels of added white space on each side of the barcode. Default value: 7.00 Minimum: 0.00 Maximum: 20.00
+    convenience init?(code128Barcode: String, quietSpace: Double = 7.0) throws {
+        if quietSpace < 0.0 || quietSpace > 20.00 {
+            throw NSError(domain: "quietSpace must between 0.0 to 20.0", code: 0, userInfo: nil)
+        }
```
Comment (guykogus) [inline Sources/SwifterSwift/UIKit/UIImageExtensions.swift:386]: Likewise, don't do this error handling here. Add an `assert` if you want, but don't make this a throwing function.

**Comment by LucianoPAlmeida** [comment]
@heart Thanks for the PR :) 
Closing due inactivity, feel free to reopen if you still want to move this forward

**Pull request closed without merging**